tempesta_fw/http_parser.c

```diff
@@ -170,20 +183,29 @@ do {									\
 /* The same as __FSM_MOVE_n(), but exactly for jumps w/o data moving. */
 #define __FSM_JMP(to)			do { goto to; } while (0)
 
-#define __FSM_MATCH_MOVE_f(alphabet, to, field)				\
+#define __FSM_MATCH_MOVE_fixup_pos(alphabet, to, field, fixup_pos)	\
 do {									\
 	__fsm_n = __data_remain(p);					\
 	__fsm_sz = tfw_match_##alphabet(p, __fsm_n);			\
 	if (unlikely(__fsm_sz == __fsm_n)) {				\
 		/* Continue field processing on next skb. */		\
 		BUG_ON(!(field)->ptr);					\
-		__msg_field_fixup(field, data + len);			\
+		if (fixup_pos)						\
+			__msg_field_fixup_pos(field, p, __fsm_sz);	\
+		else							\
+			__msg_field_fixup(field, data + len);		\
```

Comment:
I don't think this is the right change. The role of `__msg_field_fixup()` is to cut off (stop) a chunk at specified address. Same is true for `__msg_field_finish()`, just with an extra action of setting `TFW_STR_COMPLETE` flag.

Under the circumstances it appears that `p + __fsm_sz` is the same end address and `data + len`. Thus there's no need for `fixup_pos` argument, and there's no need for the whole `__FSM_MATCH_MOVE_fixup_pos()` macro.

Reply:
@keshonok Thank you for review.
The restriction of `__msg_field_fixup()` usage is that it always takes `data` pointer (i.e. the beginning of currently parsed part of message) as the beginning data chunk in `tfw_http_msg_add_str_data()`. This is suitable when we don't need to split the part of currently parsed message into separate `TfwStr` chunks (except of `TFW_POSTPONE` of course - but in this case the start `data` pointer is already set to the right place for next `TfwStr` chunk). But if we split the message via `__msg_field_fixup()` - then separate `TfwStr` chunks will point to the same `data` pointer in the real message, which is wrong behavior.  This can happen when we get to `Req_UriAbsPath` from `Req_UriMark`, and `req->uri_path` already has some part of message. So we need to explicitly specify the beginning pointer for the next chunk - via `__msg_field_fixup_pos()`.
Similar behavior we have with `__msg_hdr_chunk_fixup()`, but for headers parsing (e.g. in `__req_parse_cookie()` function).

Reply:
Done.